src/prototype_room_utils.js

```diff
@@ -186,21 +186,168 @@ Room.prototype.checkRoleToSpawn = function(role, amount, targetId, targetRoom, l
   this.memory.queue.push(creepMemory);
 };
 
-Room.prototype.getPartConfig = function(energy, parts) {
-  var sum = 0;
-  var i = 0;
-  var partConfig = [];
-  while (sum < energy && partConfig.length < 50) {
-    var part = parts[i % parts.length];
-    if (sum + BODYPART_COST[part] <= energy) {
-      partConfig.push(part);
-      sum += BODYPART_COST[part];
-      i += 1;
-    } else {
-      break;
+/**
+ * Room.prototype.checkParts use for check if a bodyPart can be add to total body.
+ *
+ * @param {Array} parts Array of body parts.
+ * @param {Number} actualCost Cost of body before function call.
+ * @param {Number} energy energy allow for spawn.
+ */
+
+Room.prototype.checkParts = function(parts, actualCost, energy) {
+  if (!parts) { return; }
+  parts.forEach(
+    function(p) {
+      actualCost += BODYPART_COST[p];
+      if (actualCost > energy) { return; }
+    }
+  );
+  return actualCost;
+};
+
+/**
+ * Room.prototype.getSettings use for return creep spawn settings
+ * adapted to room configuration
+ *
+ * @param {Collection} creep queue's creep spawn basic datas
+ */
+
+Room.prototype.getSettings = function(creep) {
+  let datas = {};
+  let i = 0;
+  let settings = roles[creep.role].settings;
+  let parts = settings.parts || {};
+  let energy = settings.energy || {};
+
+  let getKey = function(part, i) {
+    let keys = Object.keys(part);
+    let value = _.get(this, settings.param[i], 1);
+    let previousKey = 0;
+    _.forEach(keys, function(k) {
+      if (part[k] && value < k & previousKey !== 0) {
+        return previousKey;
+      }
+      previousKey = k;
+    });
+    return previousKey;
+
+  };
+  let Settings = parts;
```

Comment:
nit picking `Settings` => `settings` or other name. Could be confusion with the first `settings`